[ci] Clean up hlc test setup

tobil4sk opened the pull request:
hlc tests and jit tests are now separated and can be toggled using `--skip-hlc` and `--skip-hl-jit` flags.

Disabling the hl-jit tests on mac arm64 allows the hlc tests to run properly.

tobil4sk committed:
[ci] Separate hlc test from regular hl tests

tobil4sk committed:
[ci] Allow specifying jit vs hlc for hl tests

Controlled using the flags `--skip-hl-jit` and `--skip-hlc`

tobil4sk committed:
[ci] Allow ci to use preexisting hashlink install

tobil4sk committed:
[ci] Run hl for x86 and arm64 mac separately

yuxiaomao commented:
For the misc test, I'm wondering what happens if one misc project needs to be run with multiple hxml, and some of them are hl-jit, the others are hl-c ?

tobil4sk commented:
> For the misc test, I'm wondering what happens if one misc project needs to be run with multiple hxml, and some of them are hl-jit, the others are hl-c ?

I suppose with this setup you'd have to add the same test to both the hl and hlc folders. If they exist in the same place, the only way to check which is which is by reading the content of the hxml which the misc test runner currently doesn't support.

Is there a particular test you're thinking of adding soon that might be like this? Otherwise, maybe it's best to adjust to that if/when it ever comes up. It seems currently hl tests are for general genhl issues and hlc is for issues specific to hlc, so it's probably one or the other? Unit tests will also always run for both.

tobil4sk committed:
Merge remote-tracking branch 'upstream/development' into ci-hlc-mac-arm64

tobil4sk committed:
Merge remote-tracking branch 'upstream/development' into ci-hlc-mac-arm64

yuxiaomao commented:
I was thinking about 12039, it does not need to be run in hl/jit or hl/c, but it's good to be run for at least one of them: but it won't run if we have `--skip-hl-jit` here.
It's actually a unit test on any target who have `Single` type, but it's placed in a project to compare `$type` output.
(But I think it does not really matter because as long as we don't skip it on haxe main repo)

Otherwise, seems fine to me.

tobil4sk commented:
> (But I think it does not really matter because as long as we don't skip it on haxe main repo)

It is skipped for mac arm64, because jit is currently not supported there. However, I suppose the misc tests don't actually require execution and the hl only tests currently seem to be applicable to both hl jit and hlc.

Maybe it is best to not generate output for the misc/hl tests so it is clear the are general genhl tests, and run them even if `--skip-hl-jit` is passed?

yuxiaomao commented:
I just found the description of misc project on your comment on another PR https://github.com/HaxeFoundation/haxe/blob/development/tests/misc/README.md

> If compile.hxml was found and haxe compile.hxml finished with exit code 0 then a test is considered successful.

I think output is still needed (to call genhl/hl2c and found related error), but execution on hl-jit/hl-c doesn't seem necessary (only some check with `haxe --run` e.g. in 10184). So maybe we do not need to skip misc tests.

tobil4sk committed:
[tests] Make misc hl tests not specific to jit

By compiling the tests with --no-output, we ensure they won't generate
a .hl file, so they are general genhl tests that don't require the jit
to run.

tobil4sk commented:
By no output I meant doing something this:
```hxml
--hl _
--no-output
```

This will still print out errors from genhl, but no .hl file is created, so it ensures the test doesn't have to execute via jit. That way it's safe to run these with `--skip-hl-jit`.

For hlc, we can't use `--no-output` as the error might occur during c compilation (e.g. broken C identifier).

Does this seem ok? I have pushed an update in line with this.

Simn commented:
I don't really understand what we're doing here, but with `--no-output` in place you shouldn't see any generator-specific errors because genhl isn't involved at all.

tobil4sk committed:
[tests] Remove --no-output from misc hl tests

This means the code will not reach genhl, so genhl errors won't be seen.

tobil4sk commented:
> I don't really understand what we're doing here, but with --no-output in place you shouldn't see any generator-specific errors because genhl isn't involved at all.

Ah in that case I have misunderstood the purpose of the flag. What I want is to ensure that tests in misc/hl will still pass through genhl, but won't actually write a .hl file, which makes sure someone can't add a test there that requires actually executing hl jit.

I'll remove the flag. For now, if someone adds a misc hl test that actually attempts to run the file, it will cause an error on the mac arm64 ci.

yuxiaomao commented:
FYI: `no_output` will not run the generator
https://github.com/HaxeFoundation/haxe/blob/6cd1f655007ccdf54123ad6a87d5a98ad84e9aa7/src/compiler/compiler.ml#L419

I'm thinking about hlc misc tests too, my idea is that all the tests can still stay in the `hl` folder: If we specified `--skip-hlc`, we run the hxml without `-D hlgen.makefile` so no auto-compilation with the C compiler (we don't run the C exe in misc, and  `-D hlgen.makefile` to a hl-jit hxml won't do anything, so it's safe), and we can still check that the hl2c part is run without error.

tobil4sk commented:
I added --skip-hlc for mainly two reasons:
1. To be able to avoid hlc compile times locally when just trying to test an hl change.
2. To be able to easily check whether an hl implementation complies with the tests (e.g. hl-jit-rs, I also wanted to test -D interp recently). For this purpose hlc is not relevant, so ideally it would be possible to completely skip hlc-specific tests.

Generating hlc but not compiling it does not bring much benefit as a lot of hlc tests are ineffective without running the c compiler, since they are for issues where hl2c used to give invalid c code. Also, it doesn't cover the second use case.

> I'm thinking about hlc misc tests too, my idea is that all the tests can still stay in the hl folder

We could have hlc under `misc/hl/hlc` if the concern is about keeping them together? I think having a logical separation between tests for genhl vs hl2c would make some sense anyway.

yuxiaomao commented:
Ok, I agree with your 2nd use case, and indeed it'll make more sens to skip hlc test completely.
